### DISCUSSION TITLE
question on regrid_data_plane

### DISCUSSION THREAD

**Discussion by haksulee** [dtcenter/METplus#2778] [File I/O] [answered]
Hello

I am new to MET and METPlus, and have a question on regrid_data_plane. 
My goal is to regrid AORC data on the GEFSv12 grid.   
Example files are AORC-OWP_2017103120z.nc4 and gep05.t00z.pgrb2.0p25.f060 at the google folder with an access permission to met_help@ucar.edu: https://drive.google.com/drive/folders/1iVSAWBVb5Lsezf3A3dBZcFdDwx1Pt0EW

I downloaded docker images for met:11.1-latest and metplus:5.1-latest from https://dtcenter.org/community-code/metplus/metplus-5-1-existing-builds ,  and tried the following command inside a docker/container: 

____________________________

root@d0431ca818f7:/met/met-11.1/tutorial/regridding# regrid_data_plane AORC-OWP_2017103120z.nc4 G130 output/a.nc -field 'name="APCP_surface"; level="surface";'
DEBUG 1: Start regrid_data_plane by root(0) at 2024-11-08 14:44:13Z  cmd: regrid_data_plane AORC-OWP_2017103120z.nc4 G130 output/a.nc -field name="APCP_surface"; level="surface"; 
DEBUG 1: Reading data file: AORC-OWP_2017103120z.nc4
ERROR  : 
ERROR  : process_data_file() -> "AORC-OWP_2017103120z.nc4" not a valid data file
ERROR  : 
root@d0431ca818f7:/met/met-11.1/tutorial/regridding# 
____________________________

Based on https://mailman.ucar.edu/pipermail/met_help/2019-July/003574.html , I added file_type=NETCDF_NCCF;' -v2 which still didn't work 
____________________________
root@d0431ca818f7:/met/met-11.1/tutorial/regridding# regrid_data_plane AORC-OWP_2017103120z.nc4 G130 output/a.nc -field 'name="APCP_surface"; level="surface"; file_type=NETCDF_NCCF;' -v2
DEBUG 1: Start regrid_data_plane by root(0) at 2024-11-08 14:29:51Z  cmd: regrid_data_plane AORC-OWP_2017103120z.nc4 G130 output/a.nc -field name="APCP_surface"; level="surface"; file_type=NETCDF_NCCF; -v2 
ERROR  : 
ERROR  : CommandLine::next_option() -> unrecognized command-line switch "-v2"
ERROR  : 
root@d0431ca818f7:/met/met-11.1/tutorial/regridding# 
____________________________

Please, teach me how I can do regridding the AORC data correctly. 

Thank you
Haksu

docker ps

d0431ca818f7   dtcenter/met:11.1-latest    "/bin/bash"   23 hours ago   Up 23 hours             musing_margulis

docker image ls

dtcenter/met                11.1-latest            9b84d6d1cc66   4 months ago    3.78GB
dtcenter/metplus            5.1-latest             8ed1439300ae   15 months ago   3.93GB
dtcenter/metplus-training   latest                 52138718a387   15 months ago   2.24GB
dtcenter/metplus-data       3.1-met_tool_wrapper   1baf40d4fd28   4 years ago     1.13GB
dtcenter/met                8.1                    aa0307df830c   4 years ago     2.68GB

**Comment by JohnHalleyGotway** [comment, marked as answer]
@haksulee, thanks for sending along your sample data. Good news, this data can be easily read by the MET tools. And yes, since the global `conventions` attribute is missing from the `AORC-OWP_2017103120z.nc4` file to identify this as being a [CF-compliant NetCDF file](https://cfconventions.org/), you do need to use the `file_type=NETCDF_NCCF;` option to explicitly tell MET how to interpret it.

I do always recommend getting started with new datasets by running the Plot-Data-Plane tool. And doing so using verbosity level 4 (`-v 4`) will also print out timing information about the data being read. This step confirms that MET can read the data from your input file, place it at the right spot on earth, and the log messages confirm the metadata.

Here's a command to try:
```
plot_data_plane AORC-OWP_2017103120z.nc4 APCP_surface.ps \
name="APCP_surface"; level="(0,*,*)"; file_type=NETCDF_NCCF; -v 4
```
Here's some log output with the metadata:
```
DEBUG 4: Grid information:
DEBUG 4:    Projection: Lat/Lon
DEBUG 4:    Nx: 8401
DEBUG 4:    Ny: 4201
DEBUG 4:    lat_ll: 20.000
DEBUG 4:    lon_ll: 130.000
DEBUG 4:    delta_lat: 0.008
DEBUG 4:    delta_lon: 0.008
DEBUG 4:    wrapLon: false
DEBUG 4: 
DEBUG 4: Data plane information:
DEBUG 4:    plane min: 0
DEBUG 4:    plane max: 36.4
DEBUG 4:    valid time: 20171031_200000
DEBUG 4:    lead time: 000000
DEBUG 4:    init time: 20171031_200000
DEBUG 4:    accum time: 000000
```
And here's the resulting image:
![Screenshot 2024-11-08 at 11 32 43 AM](https://github.com/user-attachments/assets/a3865278-08c9-4504-9e50-1663b525445a)

And plotting 2m temperature with `'name="TMP_2maboveground"; level="(0,*,*)"; file_type=NETCDF_NCCF;'` makes for a bit more colorful image:
![Screenshot 2024-11-08 at 11 34 05 AM](https://github.com/user-attachments/assets/332f81ea-4da7-4dad-9b23-92309bcbe8e6)

Here's an example of running `regrid_data_plane` to regrid this data to NCEP Grid 130, using the default nearest neighbor interpolation method. See the [User's Guide](https://met.readthedocs.io/en/latest/Users_Guide/reformat_grid.html#regrid-data-plane-tool) for more details on that:
```
regrid_data_plane AORC-OWP_2017103120z.nc4 G130 APCP_surface_G130.nc \
-field 'name="APCP_surface"; level="(0,*,*)"; file_type=NETCDF_NCCF;'
```
Here's a couple more details you may find useful.
- When setting the `level` string for NetCDF variables with 3+ dimensions, you need to define a string like this `(N,*,*)` where the `*`'s go into the slots for the spatial lat/lon dimensions. The `N` value is either a 0-based integer defining the index for that dimension (e.g. `(0,*,*)`) or a string defining it's value (e.g. `(@20171031_200000,*,*)`). All of the commands listed above would work with this level string too. For files with a time dimension of length 1, just putting `0` is easy enough. But for files with many time values, you may find specifying the timestring instead (in YYYYMMDD[_HH[MMSS]] format) to be more convenient.
- Yes, you can do regridding using the `regrid_data_plane` tool, but regridding on-the-fly is also supported by many of the MET tools. For example, if you ran Grid-Stat to compare data `gep05.t00z.pgrb2.0p25.f060` forecast data to the `AORC-OWP_2017103120z.nc4` analysis field, you could set up the configuration to regrid the forecast data to the observation domain with `to_grid = OBS` or vice-versa with `to_grid = FCST`. That way you can do it during the evaluation rather than having to run 2 commands... although both options work fine.

Hope this helps get you going. Glad it was an easy one!

**Reply by haksulee** [reply]
@JohnHalleyGotway 

Hi John

Thank you very much for detailed explanation and help.  I tried below which worked without errors:
regrid_data_plane AORC-OWP_2017103120z.nc4 G130 APCP_surface_G130.nc \
-field 'name="APCP_surface"; level="(0,*,*)"; file_type=NETCDF_NCCF;'
I may try Grid-Stat later.  The problem solved.

I have a different question. 
Why separate docker images exist for met 11.1-latest and metplus 5.1-latest ? 
In which case, should I use met 11.1   or metplus 5.1-latest ? 
Can both met and metplus come in a single docker image ? 

Thank you

**Reply by JohnHalleyGotway** [reply]
Yes, the [METplus docker image](https://hub.docker.com/repository/docker/dtcenter/metplus/general) is built on top of the [MET docker image](https://hub.docker.com/repository/docker/dtcenter/met/general) and adds the METplus Python wrappers.

For example, this command shows where `plot_data_plane` can be found inside that METplus container:
```
docker run -it --rm dtcenter/metplus:5.1-latest which plot_data_plane
/usr/local/bin/plot_data_plane
```
For the next coordinated release, METplus version 6.0.0, we're also adding the METplus Analysis Tools which provide additional Python plotting functionality. Prior to the introduction of METplus wrappers, some users ran the MET tools directly. But we definitely encourage running MET through the METplus Wrappers. So I'd recommend using [dtcenter/metplus:5.1-latest](https://hub.docker.com/layers/dtcenter/metplus/5.1-latest/images/sha256-cf074e663b77e855325f0aa540666304d0212176c3582aafba2c6a4e206af23a?context=repo).

**Reply by haksulee** [reply]
Ok, thank you very much for the explanation and the help.   This discussion can be closed.
